datafusion/core/src/physical_plan/sorts/merge.rs

```diff
@@ -227,11 +239,21 @@ impl<C: Cursor> SortPreservingMergeStream<C> {
             if self.advance(stream_idx) {
                 self.loser_tree_adjusted = false;
                 self.in_progress.push_row(stream_idx);
-                if self.in_progress.len() < self.batch_size {
+
+                // stop sorting if fetch has been reached
+                if self
```

Comment:
Nit: this might be easier to read if you moved the "fetch_reached" into its own function

like 

```rust
if self.fetch_reached() {
  self.aborted = true;
} else { ..
```